animation_nodes/nodes/spline/splines_from_edges_utils.pyx

```diff
@@ -0,0 +1,113 @@
+from ... data_structures cimport (
+    Mesh, Vector3DList, EdgeIndicesList, PolySpline,
+    VirtualDoubleList, FloatList, IntegerList
+)
+
+def splinesFromBranches(Vector3DList vertices, EdgeIndicesList edges, VirtualDoubleList radii):
+    cdef int i, j
+    cdef int edgesAmount = edges.length
+    cdef int verticesAmount = vertices.length
+
+    # Compute how many neighbour each vertex have.
+    cdef IntegerList neighboursAmounts = IntegerList.fromValue(0, length = verticesAmount)
+    for i in range(edgesAmount):
+        neighboursAmounts.data[edges.data[i].v1] += 1
+        neighboursAmounts.data[edges.data[i].v2] += 1
+
+    # Compute the start index of each group of neighbours of each vertex.
+    cdef IntegerList neighboursStarts = IntegerList(length = verticesAmount)
+    cdef int start = 0
+    for i in range(verticesAmount):
+        neighboursStarts.data[i] = start
+        start += neighboursAmounts.data[i]
+
+    # Keep track of how many index is in each group of neighbours at each iteration.
+    cdef IntegerList filledSpaces = IntegerList.fromValue(0, length = verticesAmount)
+
+    # Compute the indices of neighbouring vertices of each vertex.
+    cdef IntegerList neighbours = IntegerList(length = edgesAmount * 2)
+    for i in range(edgesAmount):
+        neighbours.data[neighboursStarts.data[edges.data[i].v1] +
+        filledSpaces.data[edges.data[i].v1]] = edges.data[i].v2
+        filledSpaces.data[edges.data[i].v1] += 1
+
+        neighbours.data[neighboursStarts.data[edges.data[i].v2] +
+        filledSpaces.data[edges.data[i].v2]] = edges.data[i].v1
+        filledSpaces.data[edges.data[i].v2] += 1
+
+    # Find the indices of the vertices that are not connected to two vertices.
+    cdef IntegerList nonBipolarVertices = IntegerList(length = verticesAmount)
+    cdef int nonBipolarVertsCount = 0
+    for i in range(verticesAmount):
+        if neighboursAmounts.data[i] != 2:
+            nonBipolarVertices.data[nonBipolarVertsCount] = i
+            nonBipolarVertsCount += 1
+    nonBipolarVertices.length = nonBipolarVertsCount
+    nonBipolarVertices.shrinkToLength()
+
+    # Generate Splines.
+    cdef int count
+    cdef list splines = []
+    cdef PolySpline spline
+    cdef FloatList splineRadii
+    cdef Vector3DList splineVertices
+    cdef int nonBipolarVertex, currentVertex, nextVertex
+    for i in range(nonBipolarVertsCount):
+        nonBipolarVertex = nonBipolarVertices.data[i]
+        for j in range(neighboursAmounts.data[nonBipolarVertex]):
+            nextVertex = neighbours.data[neighboursStarts.data[nonBipolarVertex] + j]
+            if (filledSpaces.data[nextVertex] == neighboursAmounts.data[nextVertex] or
+                                                neighboursAmounts.data[nextVertex] != 2):
+                splineVertices = Vector3DList.__new__(Vector3DList, length = verticesAmount)
```

Comment:
these lists are wayy to large in many cases. Better use the `append_LowLevel` function instead of allocating huge arrays for every spline.

Reply:
What should the initial capacity of those lists be?

Reply:
Just don't give them an initial capacity, or maybe `2`.